File: android/src/main/java/com/rebeloid/unity_ads/UnityAdsShowListener.java
package com.rebeloid.unity_ads;

import com.unity3d.ads.IUnityAdsShowListener;
import com.unity3d.ads.UnityAds;

public class UnityAdsShowListener implements IUnityAdsShowListener {
    private final PlacementChannelManager placementChannelManager;

    public UnityAdsShowListener(PlacementChannelManager placementChannelManager) {
        this.placementChannelManager = placementChannelManager;
    }

    @Override
    public void onUnityAdsShowFailure(String placementId, UnityAds.UnityAdsShowError error, String message) {
        placementChannelManager.invokeMethod(UnityAdsConstants.SHOW_FAILED_METHOD, placementId, convertError(error), message);
    }

    @Override
    public void onUnityAdsShowStart(String placementId) {
        placementChannelManager.invokeMethod(UnityAdsConstants.SHOW_START_METHOD, placementId);
    }

    @Override
    public void onUnityAdsShowClick(String placementId) {
        placementChannelManager.invokeMethod(UnityAdsConstants.SHOW_CLICK_METHOD, placementId);
    }

    @Override
    public void onUnityAdsShowComplete(String placementId, UnityAds.UnityAdsShowCompletionState state) {
        if (state == UnityAds.UnityAdsShowCompletionState.SKIPPED) {
            placementChannelManager.invokeMethod(UnityAdsConstants.SHOW_SKIPPED_METHOD, placementId);
        } else if (state == UnityAds.UnityAdsShowCompletionState.COMPLETED) {
            placementChannelManager.invokeMethod(UnityAdsConstants.SHOW_COMPLETE_METHOD, placementId);
        }
    }

    private String convertError(UnityAds.UnityAdsShowError error) {
        switch (error) {
            case NOT_INITIALIZED:
                return "notInitialized";
            case NOT_READY:
                return "notReady";
            case VIDEO_PLAYER_ERROR:
                return "videoPlayerError";
            case INVALID_ARGUMENT:
                return "invalidArgument";
            case NO_CONNECTION:
                return "noConnection";
            case ALREADY_SHOWING:
                return "alreadyShowing";
            case INTERNAL_ERROR:
                return "internalError";
            case TIMEOUT:
                return "timeout";
            default:
                return "";
        }
    }
}

File: ios/Classes/UnityAdsShowListener.swift
import Flutter
import UnityAds

public class UnityAdsShowListener : NSObject, UnityAdsShowDelegate {
    let placementChannelManager: PlacementChannelManager
    
    init(placementChannelManager: PlacementChannelManager) {
        self.placementChannelManager = placementChannelManager
    }
    
    public func unityAdsShowComplete(_ placementId: String, withFinish state: UnityAdsShowCompletionState) {
        if (state == UnityAdsShowCompletionState.showCompletionStateSkipped) {
            placementChannelManager.invokeMethod(UnityAdsConstants.SHOW_SKIPPED_METHOD, placementId)
        } else if (state == UnityAdsShowCompletionState.showCompletionStateCompleted) {
            placementChannelManager.invokeMethod(UnityAdsConstants.SHOW_COMPLETE_METHOD, placementId)
        }
    }

    public func unityAdsShowFailed(_ placementId: String, withError error: UnityAdsShowError, withMessage message: String) {
        placementChannelManager.invokeMethod(UnityAdsConstants.SHOW_FAILED_METHOD, placementId, withErrorCode: convertError(error), withMessage: message)
    }
    
    private func convertError(_ error: UnityAdsShowError) -> String {
        switch (error) {
        case .showErrorNotInitialized:
            return "notInitialized"
        case .showErrorNotReady:
            return "notReady"
        case .showErrorVideoPlayerError:
            return "videoPlayerError"
        case .showErrorInvalidArgument:
            return "invalidArgument"
        case .showErrorNoConnection:
            return "noConnection"
        case .showErrorAlreadyShowing:
            return "alreadyShowing"
        case .showErrorInternalError:
            return "internalError"
        case .showErrorTimeout:
            return "timeout"
        @unknown default:
            return ""
        }
    }
    
    public func unityAdsShowStart(_ placementId: String) {
        placementChannelManager.invokeMethod(UnityAdsConstants.SHOW_START_METHOD, placementId)
    }
    
    public func unityAdsShowClick(_ placementId: String) {
        placementChannelManager.invokeMethod(UnityAdsConstants.SHOW_CLICK_METHOD, placementId)
    }

}
